slmctl/slmctl.8
.\" $Id$
.\" %GPL_START_LICENSE%
.\" ---------------------------------------------------------------------
.\" Copyright 2008-2018, Pittsburgh Supercomputing Center
.\" All rights reserved.
.\"
.\" This program is free software; you can redistribute it and/or modify
.\" it under the terms of the GNU General Public License as published by
.\" the Free Software Foundation; either version 2 of the License, or (at
.\" your option) any later version.
.\"
.\" This program is distributed WITHOUT ANY WARRANTY; without even the
.\" implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR
.\" PURPOSE.  See the GNU General Public License contained in the file
.\" `COPYING-GPL' at the top of this distribution or at
.\" https://www.gnu.org/licenses/gpl-2.0.html for more details.
.\" ---------------------------------------------------------------------
.\" %END_LICENSE%
.\" %PFL_MODULES ctl rpc %
.Dd April 19, 2016
.Dt SLMCTL 8
.ds volume PSC \- SLASH2 Administrator's Manual
.Os http://www.psc.edu/
.Sh NAME
.Nm slmctl
.Nd
.Tn SLASH2 MDS
daemon runtime control
.Sh SYNOPSIS
.Nm slmctl
.Op Fl HInV
.Op Fl p Ar paramspec
.Op Fl S Ar socket
.Op Fl s Ar showspec
.Op Ar cmd arg ...
.Sh DESCRIPTION
The
.Nm
utility examines and manipulates
.Xr slashd 8
behavior.
.Pp
.\" %PFL_INCLUDE $PFL_BASE/doc/pflctl/cmd.mdoc {
.\"	cmds => {
.\"		stop => <<'EOF'
.\"			Instruct
.\"			.Xr slashd 8
.\"			to completely terminate daemon execution.
.\"		EOF
.\"	}
The supported commands are as follows:
.Bl -tag -width 3n
.It Cm stop
Instruct
.Xr slashd 8
to completely terminate daemon execution.
.El
.\" }%
.Pp
The options are as follows:
.Bl -tag -width 3n
.\" %PFL_INCLUDE $PFL_BASE/doc/pflctl/H.mdoc {
.It Fl H
Suppress headers when showing data displays.
.\" }%
.\" %PFL_INCLUDE $PFL_BASE/doc/pflctl/I.mdoc {
.It Fl I
Display large numerical values accurately instead of rounded with
postfixes of demonination such as
.Sq G
for gibibytes,
.Sq M
for mebibytes, etc.\&
.Pq Do inhuman Dc number mode .
.\" }%
.\" %PFL_INCLUDE $PFL_BASE/doc/pflctl/n.mdoc {
.It Fl n
Suppress
.Tn DNS
lookups for network addresses.
.\" }%
.\" %PFL_INCLUDE $PFL_BASE/doc/pflctl/p.mdoc {
.\"	log_xr => "in\n.Xr slashd 8\n",
.\"	params => {
.\"		'pid' => "Daemon system process ID.",
.\"		"sys.next_fid" => "Next file identifier\n.Pq Tn FID\nthat will be used for new file creation.",
.\"		"sys.global" => "Boolean switch to enable the global mount feature.",
.\"		'sys.nbrq_outstanding'
.\"		     => "Number of currently outstanding asynchronous RPCs.",
.\"		"sys.resources" => <<EOF .
.\"			Settings and fields specific to network peers.
.\"			.Bl -tag -width 13n -offset 3n
.\"			EOF
.\"			build_list(
.\"				xid => <<EOF,
.\"					Highest journal transaction identifier peer has received
.\"					for either namespace metadata updates or garbage
.\"					reclamation updates.
.\"				EOF
.\"				disable_bia => <<EOF,
.\"					Whether bmap write lease assignments are administratively
.\"					disabled
.\"					.Pq ION only .
.\"					EOF
.\"				disable_gc => <<EOF,
.\"					Whether garbage collection and reclamation is administratively
.\"					disabled
.\"					.Pq ION only .
.\"					EOF
.\"			) . ".El",
.\"	}
.It Fl p Ar paramspec
Query/manipulate operational parameters.
.Ar paramspec
has the following format:
.Pp
.Bd -unfilled -offset 3n
.Sm off
.Oo Ar thread-name Ns Li .\& Oc Ar param
.Op Oo Li +- Oc Li = Ar value
.Sm on
.Ed
.Pp
Some parameters may be read- or write-only.
Some support modification by the assignment operators
.Li +=
and
.Li -= .
.Pp
See
.Sx Thread Specification
for details on specifying
.Ar thread-name .
The default behavior is to apply the operation to all applicable threads.
.Pp
.Ar param
may be one of the following:
.Bl -tag -width 1n -offset 3n
.It Cm lnet.networks
.Tn LNET
network configuration.
.It Cm lnet.port
If applicable,
.Xr tcp 7
port to use to connect to remote
.Tn LNET
peers.
.It Cm log.file
File path name where log messages will be written.
This field is write-only.
If the
.Li +=
assignment operator is used, this file will not be truncated.
.It Cm log.format
The header prepended to server log messages.
See the description of
.Ev PSC_LOG_FORMAT
in
.Xr slashd 8
for details on this format.
.It Cm log.level Ns Op . Ns Ar subsystem
The logging level of debug message output.
.Pp
See the description of
.Ev PSC_LOG_LEVEL_ Ns Ar subsys
in
.Xr slashd 8
for a list of available subsystems.
If
.Ar subsystem
is not specified, all subsystems will be accessed.
.Pp
See the description of
.Ev PSC_LOG_LEVEL
in
.Xr slashd 8
for recognized log level values.
.It Cm log.points
List of places in the source code (in the form
.Dq file:line-number )
where granular logging can be enabled on a case-by-case
basis.
Usually assigned to by the
.Li +=
operator.
.It Cm pid
Daemon system process ID.
.It Cm pool. Ns Ar name
Access the memory pool specified by
.Ar name .
The following sub-fields are available:
.Pp
.Bl -tag -compact -offset 3n -width 13n
.It Cm free
Current number of allocated entries available for use in pool.
.It Cm max
Upper bound for number of entries to which auto-sized pools can grow.
.It Cm min
Lower bound for number of entries to which auto-sized pools can shrink.
.It Cm reap
Assign a value to reap the specified number of entries.
Only applies to pools that have a reaper routine.
This field is write-only.
.It Cm thres
Threshold for unused items for auto-sized pools before items are freed.
.It Cm total
Current number of entries contained in pool.
.El
.It Cm rlim
Process resource limits.
See
.Xr getrlimit 2
or
.Xr ulimit 1
for more information.
.Pp
.Bl -column "msgqueue" -offset 3n
.It Sy Name Ta Sy Field
.It =============================
.It Cm cpu        Ta Dv RLIMIT_CPU
.It Cm csize      Ta Dv RLIMIT_CORE
.It Cm dsize      Ta Dv RLIMIT_DATA
.It Cm fsize      Ta Dv RLIMIT_FSIZE
.It Cm locks      Ta Dv RLIMIT_LOCKS
.It Cm maxproc    Ta Dv RLIMIT_NPROC
.It Cm mem        Ta Dv RLIMIT_AS
.It Cm mlock      Ta Dv RLIMIT_MEMLOCK
.It Cm msgqueue   Ta Dv RLIMIT_MSGQUEUE
.It Cm nice       Ta Dv RLIMIT_NICE
.It Cm nofile     Ta Dv RLIMIT_NOFILE
.It Cm rss        Ta Dv RLIMIT_RSS
.It Cm rtprio     Ta Dv RLIMIT_RTPRIO
.It Cm rttime     Ta Dv RLIMIT_RTTIME
.It Cm sigpndg    Ta Dv RLIMIT_SIGPENDING
.It Cm stksize    Ta Dv RLIMIT_STACK
.El
.It Cm rusage
Process resource usage information.
See
.Xr getrusage 2 .
.It Cm sys.global
Boolean switch to enable the global mount feature.
.It Cm sys.nbrq_outstanding
Number of currently outstanding asynchronous RPCs.
.It Cm sys.next_fid
Next file identifier
.Pq Tn FID
that will be used for new file creation.
.It Cm sys.resources
Settings and fields specific to network peers.
.Bl -tag -width 13n -offset 3n
.It Cm disable_bia
Whether bmap write lease assignments are administratively
disabled
.Pq ION only .
.It Cm disable_gc
Whether garbage collection and reclamation is administratively
disabled
.Pq ION only .
.It Cm xid
Highest journal transaction identifier peer has received
for either namespace metadata updates or garbage
reclamation updates.
.El
.El
.\" }%
.\" %PFL_INCLUDE $PFL_BASE/doc/pflctl/S.mdoc {
.\"	daemon	=> qq{slashd},
.\"	sock	=> "/var/run/slashd. Ns Ic %h Ns Pa .sock"
.It Fl S Ar socket
Specify an alternative socket file.
The following tokens are replaced in the file name specified:
.Pp
.Bl -tag -offset 3n -width Ds -compact
.It Cm %h
the machine hostname
.It Cm %n
the daemon executable base name, i.e.\&
.Dq slashd
.It Cm %%
a literal
.Sq %
character
.El
.Pp
The default is
.Pa /var/run/slashd. Ns Ic %h Ns Pa .sock .
.\" }%
.\" %PFL_INCLUDE $PFL_BASE/doc/pflctl/show.mdoc {
.\"	show => {
.\"		bmap		=> qq{In-memory bmaps},
.\"		bml		=> qq{Outstanding bmap leases.},
.\"		connections	=> <<EOF,
.\"			Status of peer nodes on the SLASH2 deployment.
.\"			.Pp
.\"			The following flag values may appear:
.\"			.Pp
.\"			.Bl -column "Flagxx" -offset 3n
.\"			.It Sy Flag Ta Sy Description
.\"			.It =====================================
.\"			.It Li C Ta Connecting
.\"			.It Li O Ta Connected
.\"			.It Li F Ta Freeing
.\"			.It Li P Ta Registered for keep alive (ping)
.\"			.El
.\"			EOF
.\"		fidcache	=> qq{.Tn FID\n.Pq file- Ns Tn ID\ncache members.},
.\"		odtables	=> qq{Disk-backed data files.},
.\"		repl		=> qq{Replica endpoint traffic statistics.},
.\"		statfs		=> qq{.Tn I/O\nnode backing file system statistics.},
.\"	},
.\"	hashtables => {
.\"		fidc		=> qq{files\n.Po file\n.Tn ID\ncache\n.Pc},
.\"		resnid		=> qq{network resources\n.Pq network Tn ID},
.\"	},
.\"	iostats => {
.\"		qq{jrnlrd- Ns Ar fn ,} => "",
.\"		qq{jrnlwr- Ns Ar fn} =>
.\"		    qq{Data read/written to journal disk file\n.Ar fn .\n.Pp},
.\"	},
.\"	meters => {
.\"		qq{nsupd- Ns Ar peer}	=> qq{Namespace updates to\n.Tn MDS\npeers.},
.\"		qq{reclaim- Ns Ar peer}	=> qq{Garbage collection updates to\n.Tn IO\nsystems.},
.\"	},
.\"	journals => {
.\"		qq{op-journal}	=> qq{Pending operation journal},
.\"	},
.\"	listcaches => {
.\"		fcmhbusy	=> "Files with pending activity e.g.\\&\n.Tn I/O",
.\"		fcmhidle	=> "Clean\n.Pq reapable\nfiles",
.\"		inflightbml	=> "Active bmap lease mode updates",
.\"		pendingbml	=> "Pending bmap lease mode updates",
.\"		replstwkq	=> "Replication status work buffers",
.\"	},
.\"	pools => {
.\"		bmap		=> qq{Block map structures},
.\"	}
.It Fl s Ar showspec
Show values.
.Ar showspec
has the following format:
.Bd -unfilled -offset 3n
.Sm off
.Ar param
.Op : Ar subspec
.Sm on
.Ed
.Pp
.Ar param
may be specified as any non-ambiguous prefix abbreviation of the
following:
.Pp
.Bl -tag -width 1n -offset 3n
.It Cm bmap
In-memory bmaps
.It Cm bml
Outstanding bmap leases.
.It Cm connections
Status of peer nodes on the SLASH2 deployment.
.Pp
The following flag values may appear:
.Pp
.Bl -column "Flagxx" -offset 3n
.It Sy Flag Ta Sy Description
.It =====================================
.It Li C Ta Connecting
.It Li O Ta Connected
.It Li F Ta Freeing
.It Li P Ta Registered for keep alive (ping)
.El
.It Cm fidcache
.Tn FID
.Pq file- Ns Tn ID
cache members.
.It Cm hashtables
Hash table statistics.
.Ar subspec
has the following format:
.Bd -unfilled -offset 3n
.Ar hash-table Ns Op , Ns Ar ...
.Ed
.Pp
.Ar hash-table
may be one of the following:
.Pp
.Bl -tag -compact -offset 3n -width 13n
.It Cm fidc
files
.Po file
.Tn ID
cache
.Pc
.It Cm resnid
network resources
.Pq network Tn ID
.El
.Pp
If
.Ar subspec
is left unspecified, all hash tables will be accessed.
.It Cm journals
Journal statistics.
.Ar subspec
has the following format:
.Pp
.Bd -unfilled -offset 3n
.Ar journal Ns Op , Ns Ar ...
.Ed
.Pp
.Ar journal
may be one of the following:
.Pp
.Bl -tag -compact -offset 3n -width 13n
.It Cm op-journal
Pending operation journal
.El
.Pp
If
.Ar subspec
is left unspecified, all journals will be accessed.
.It Cm listcaches
List cache statistics.
.Ar subspec
has the following format:
.Pp
.Bd -unfilled -offset 3n
.Ar list Ns Op , Ns Ar ...
.Ed
.Pp
.Ar list
may be one of the following:
.Pp
.Bl -tag -compact -offset 3n -width 13n
.It Cm fcmhbusy
Files with pending activity e.g.\&
.Tn I/O
.It Cm fcmhidle
Clean
.Pq reapable
files
.It Cm inflightbml
Active bmap lease mode updates
.It Cm pendingbml
Pending bmap lease mode updates
.It Cm replstwkq
Replication status work buffers
.El
.Pp
If
.Ar subspec
is left unspecified, all list caches will be accessed.
.It Cm lnetif
Lustre network interfaces.
.It Cm loglevels
Thread logging levels.
.Ar subspec
has the following format:
.Bd -unfilled -offset 3n
.Ar thread Ns Op , Ns Ar ...
.Ed
.Pp
See
.Sx Thread Specification
for details on specifying
.Ar thread .
If
.Ar subspec
is left unspecified, all threads will be accessed.
.It Cm meters
Report on ongoing operations in progress meters.
.Ar subspec
has the following format:
.Bd -unfilled -offset 3n
.Ar meter Ns Op , Ns Ar ...
.Ed
.Pp
.Ar meter
may be one of the following:
.Pp
.Bl -tag -compact -offset 3n -width 13n
.It Cm nsupd- Ns Ar peer
Namespace updates to
.Tn MDS
peers.
.It Cm reclaim- Ns Ar peer
Garbage collection updates to
.Tn IO
systems.
.El
.Pp
If
.Ar subspec
is left unspecified, all ongoing operations will be reported.
.It Cm odtables
Disk-backed data files.
.It Cm opstats
Operation counters/statistics.
.Ar subspec
has the following format:
.Pp
.Bd -unfilled -offset 3n
.Ar opstats Ns Op , Ns Ar ...
.Ed
.Pp
.Ar opstats
may be one of the following:
.Pp
.Bl -tag -compact -offset 3n -width 3n
.It Cm lni-rcv- Ns Ar if ,
.It Cm lni-snd- Ns Ar if
Data sent/received per
.Tn LNET
networking interface.
.Pp
.It Cm lusklnd- Ns Ar mode Ns Cm -rcv ,
.It Cm lusklnd- Ns Ar mode Ns Cm -snd
Data sent/received over userland socket networking device.
.Ar mode
may be
.Cm pasv
.Pq passive
or
.Cm aggr
.Pq aggregate .
.Pp
.It Cm rpc- Ns Ar addr Ns Cm -rcv ,
.It Cm rpc- Ns Ar addr Ns Cm -snd
Data sent/received per
.Tn RPC
peer.
.Pp
.El
.Pp
If
.Ar subspec
is left unspecified, all
.Tn I/O
statistics will be accessed.
.It Cm pools
Memory pool statistics.
.Ar subspec
has the following format:
.Bd -unfilled -offset 3n
.Ar pool Ns Op , Ns Ar ...
.Ed
.Pp
.Ar pool
may be one of the following:
.Pp
.Bl -tag -compact -offset 3n -width 13n
.It Cm bmap
Block map structures
.El
.Pp
If
.Ar subspec
is left unspecified, all pools will be accessed.
.It Cm repl
Replica endpoint traffic statistics.
.It Cm rpcrqs
Remote procedure calls (RPC).
.It Cm rpcsvcs
.Tn RPC
services.
.It Cm statfs
.Tn I/O
node backing file system statistics.
.It Cm threads
Daemon thread activity and statistics.
.Ar subspec
has the following format:
.Bd -unfilled -offset 3n
.Ar thread Ns Op , Ns Ar ...
.Ed
.Pp
See
.Sx Thread Specification
for details on specifying
.Ar thread .
If
.Ar subspec
is left unspecified, all threads will be accessed.
.It Cm workrqs
Work requests.
.El
.Pp
The special value
.Sq \&?
may also be specified to display a list of recognized values.
.\" }%
.It Fl V
Show version.
.El
.\" %PFL_INCLUDE $PFL_BASE/doc/pflctl/thr.mdoc {
.\"	thrs => {
.\"		"slmbchrqthr"			=> "Batch\n.Tn RPC\ntimeout monitor",
.\"		"slmbkdbthr"			=> "",
.\"		"slmbmaptimeothr"		=> "Bmap lease timeout monitor",
.\"		"slmconnthr"			=> "Peer resource connection monitor",
.\"		"slmctlacthr"			=> ".Nm\nconnection acceptor",
.\"		"slmctlthr"			=> ".Nm\nconnection processor",
.\"		"slmdbwkthr"			=> "",
.\"		"slmjcursorthr"			=> "Journal cursor updater thread",
.\"		"slmjnsthr"			=> "Peer\n.Pq Tn MDS\nnamespace updater",
.\"		"slmjreclaimthr"		=> "Peer\n.Pq Tn IOS\ngarbage collection notifier",
.\"		"slmjthr"			=> "Master journal thread",
.\"		"slmlnacthr- Ns Ar %s"		=> ".Tn LNET\nnetwork acceptor thread",
.\"		"slmnbrqthr"			=> "Non-blocking\n.Tn RPC\nreply handler",
.\"		"slmopstimerthr"		=> "Internal operation count updater",
.\"		"slmrcmthr"			=> "Client\n.Tn RPC\nrequest initiator thread",
.\"		"slmrmcthr Ns Ar %02d"		=> "Client\n.Tn RPC\nrequest service thread",
.\"		"slmrmithr Ns Ar %02d"		=> ".Tn I/O\nnode\n.Tn RPC\nrequest service thread",
.\"		"slmrmmthr Ns Ar %02d"		=> ".No Inter- Ns MDS RPC\nrequest service thread",
.\"		"slmupschedthr"			=> "Peer update scheduler thread",
.\"		"slmusklndplthr Ns Ar %d"	=> ".Tn LNET\nuserland socket poll thread",
.\"		"slmwkthr Ns Ar %d"		=> "Generic worker",
.\"		"slmzfskstatmthr"		=> ".Tn ZFS\nstat mount thread",
.\"	}
.Ss Thread Specification
Options which take
.Ar thread-name
parameters may be specified by one or more of the following tokens,
separated by commas:
.Pp
.Bl -tag -compact -offset 3n -width 16n
.It Cm slmbchrqthr
Batch
.Tn RPC
timeout monitor
.It Cm slmbkdbthr
.It Cm slmbmaptimeothr
Bmap lease timeout monitor
.It Cm slmconnthr
Peer resource connection monitor
.It Cm slmctlacthr
.Nm
connection acceptor
.It Cm slmctlthr
.Nm
connection processor
.It Cm slmdbwkthr
.It Cm slmjcursorthr
Journal cursor updater thread
.It Cm slmjnsthr
Peer
.Pq Tn MDS
namespace updater
.It Cm slmjreclaimthr
Peer
.Pq Tn IOS
garbage collection notifier
.It Cm slmjthr
Master journal thread
.It Cm slmlnacthr- Ns Ar %s
.Tn LNET
network acceptor thread
.It Cm slmnbrqthr
Non-blocking
.Tn RPC
reply handler
.It Cm slmopstimerthr
Internal operation count updater
.It Cm slmrcmthr
Client
.Tn RPC
request initiator thread
.It Cm slmrmcthr Ns Ar %02d
Client
.Tn RPC
request service thread
.It Cm slmrmithr Ns Ar %02d
.Tn I/O
node
.Tn RPC
request service thread
.It Cm slmrmmthr Ns Ar %02d
.No Inter- Ns MDS RPC
request service thread
.It Cm slmupschedthr
Peer update scheduler thread
.It Cm slmusklndplthr Ns Ar %d
.Tn LNET
userland socket poll thread
.It Cm slmwkthr Ns Ar %d
Generic worker
.It Cm slmzfskstatmthr
.Tn ZFS
stat mount thread
.It Cm everyone
All threads
.Pq default, where applicable
.El
.\" }%
.\" %PFL_INCLUDE $PFL_BASE/doc/env.mdoc {
.Sh ENVIRONMENT
.Bl -tag -width 3n
.It Ev CTL_SOCK_FILE
Override the default control socket file path.
.El
.\" }%
.Sh FILES
.Bl -tag -width Pa
.It Xo
.Pa /var/run/slashd. Ns Ic %h Ns Pa .sock
.Xc
default
.Xr slashd 8
control socket
.El
.Sh SEE ALSO
.Xr sladm 7 ,
.Xr slashd 8 ,
.Xr slictl 8


slmctl/slmctl.c
/* $Id$ */
/*
 * %GPL_START_LICENSE%
 * ---------------------------------------------------------------------
 * Copyright 2007-2018, Pittsburgh Supercomputing Center
 * All rights reserved.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or (at
 * your option) any later version.
 *
 * This program is distributed WITHOUT ANY WARRANTY; without even the
 * implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR
 * PURPOSE.  See the GNU General Public License contained in the file
 * `COPYING-GPL' at the top of this distribution or at
 * https://www.gnu.org/licenses/gpl-2.0.html for more details.
 * ---------------------------------------------------------------------
 * %END_LICENSE%
 */

#include <curses.h>
#include <err.h>
#include <stdio.h>
#include <stdlib.h>
#include <term.h>
#include <unistd.h>

#include "pfl/cdefs.h"
#include "pfl/ctl.h"
#include "pfl/ctlcli.h"
#include "pfl/fmt.h"
#include "pfl/log.h"
#include "pfl/pfl.h"

#include "ctl.h"
#include "ctlcli.h"
#include "pathnames.h"
#include "slashrpc.h"

#include "slashd/bmap_mds.h"
#include "slashd/ctl_mds.h"
#include "slashd/repl_mds.h"

void
packshow_bmaps(__unusedx char *bmaps)
{
	psc_ctlmsg_push(SLMCMT_GETBMAP, sizeof(struct slctlmsg_bmap));
}

void
packshow_conns(__unusedx char *conn)
{
	psc_ctlmsg_push(SLMCMT_GETCONNS, sizeof(struct slctlmsg_conn));
}

void
packshow_fcmhs(char *fid)
{
	struct slctlmsg_fcmh *scf;

	scf = psc_ctlmsg_push(SLMCMT_GETFCMHS, sizeof(*scf));
	scf->scf_fg.fg_fid = FID_ANY;

	if (fid) {
		if (strcmp(fid, "busy") == 0)
			scf->scf_fg.fg_gen = SLCTL_FCL_BUSY;
		else
			psc_fatalx("unrecognized fidcache class: %s",
			    fid);
	}
}

void
packshow_replqueued(char *res)
{
	struct slmctlmsg_replqueued *scrq;

	scrq = psc_ctlmsg_push(SLMCMT_GETREPLQUEUED, sizeof(*scrq));
}

void
packshow_statfs(__unusedx char *s)
{
	psc_ctlmsg_push(SLMCMT_GETSTATFS,
	    sizeof(struct slmctlmsg_statfs));
}

void
packshow_bml(__unusedx char *s)
{
	psc_ctlmsg_push(SLMCMT_GETBML, sizeof(struct slmctlmsg_bml));
}

int
slm_replqueued_prhdr(__unusedx struct psc_ctlmsghdr *mh,
    __unusedx const void *m)
{
	printf("%-32s %15s %15s %15s %15s\n",
	    "resource", "ingress-pending", "egress-pending", "ingress-aggr", "egress-aggr");
	return(PSC_CTL_DISPLAY_WIDTH+16);
}

void
slm_replqueued_prdat(__unusedx const struct psc_ctlmsghdr *mh,
    const void *m)
{
	int base10 = 1;
	const struct slmctlmsg_replqueued *scrq = m;

	printf("%-32s ", scrq->scrq_resname);

	base10 = (scrq->scrq_repl_ingress_pending <= 1024*1024) ? 1 : 0;
	psc_ctl_prnumber(base10, scrq->scrq_repl_ingress_pending, 15, " ");

	base10 = (scrq->scrq_repl_egress_pending <= 1024*1024) ? 1 : 0;
	psc_ctl_prnumber(base10, scrq->scrq_repl_egress_pending, 15, " ");

	base10 = (scrq->scrq_repl_ingress_aggr<= 1024*1024) ? 1 : 0;
	psc_ctl_prnumber(base10, scrq->scrq_repl_ingress_aggr, 15, " ");

	base10 = (scrq->scrq_repl_egress_aggr <= 1024*1024) ? 1 : 0;
	psc_ctl_prnumber(base10, scrq->scrq_repl_egress_aggr, 15, " ");
	printf("\n");
}

int
slm_statfs_prhdr(__unusedx struct psc_ctlmsghdr *mh,
    __unusedx const void *m)
{
	printf("%-27s %4s  %8s %7s %7s %10s  %-17s\n",
	    "resource", "flag", "capacity", "used", "remain", "utilization", "type");
	return(PSC_CTL_DISPLAY_WIDTH - 4);
}

void
slm_statfs_prdat(__unusedx const struct psc_ctlmsghdr *mh, const void *m)
{
	struct class {
		unsigned frac;
		int	 col;
	} *c, classes[] = {
		{ 85, COLOR_RED },
		{ 60, COLOR_YELLOW },
		{  0, COLOR_GREEN }
	};
	const struct slmctlmsg_statfs *scsf = m;
	const struct srt_statfs *b = &scsf->scsf_ssfb;
	char cbuf[PSCFMT_RATIO_BUFSIZ];
	char name[RES_NAME_MAX];
	int j, col = 0;

	if (b->sf_blocks)
		for (c = classes, j = 0; j < nitems(classes); j++, c++)
			if (100 * (b->sf_blocks - b->sf_bavail) /
			     b->sf_blocks >= c->frac) {
				col = c->col;
				break;
			}
	strlcpy(name, scsf->scsf_resname, sizeof(name));
	printf("%-27s %c%c%c- ", name,
	    scsf->scsf_flags & SIF_DISABLE_LEASE ?    'W' : '-',
	    scsf->scsf_flags & SIF_DISABLE_ADVLEASE ? 'X' : '-',
	    scsf->scsf_flags & SIF_DISABLE_GC ?       'G' : '-');

	printf("  ");
	/*
	 * The following uses the formula from df.c in GNU coreutils.
	 * However, we don't do integer arithmetic.
	 */
	b->sf_blocks ? psc_ctl_prnumber(0, b->sf_blocks * b->sf_frsize,
	    0, "") : printf("%7s", "-");
	printf(" ");
	b->sf_blocks ? psc_ctl_prnumber(0, (b->sf_blocks - b->sf_bfree) *
	    b->sf_frsize, 0, "") : printf("%7s", "-");
	printf(" ");
	b->sf_blocks ? psc_ctl_prnumber(0, b->sf_bavail * b->sf_frsize,
	    0, "") : printf("%7s", "-");
	printf(" ");

	setcolor(col);
	if (b->sf_blocks)
		pfl_fmt_ratio(cbuf, b->sf_blocks - b->sf_bfree,
		    b->sf_blocks - b->sf_bfree + b->sf_bavail);
	else
		strlcpy(cbuf, "-", sizeof(cbuf));
	printf("%11s", cbuf);
	uncolor();
	printf("  %-17s\n", b->sf_type);
}

void
slmctlcmd_stop(int ac, __unusedx char *av[])
{
	if (ac > 1)
		errx(1, "stop: unknown arguments");
	psc_ctlmsg_push(SLMCMT_STOP, 0);
}

void
slmctlcmd_upsch_query(int ac, char *av[])
{
	struct slmctlmsg_upsch_query *scuq;
	size_t len;

	if (ac != 1)
		errx(1, "upsch_query: no query specified");
	len = strlen(av[0]) + 1;
	scuq = psc_ctlmsg_push(SLMCMT_UPSCH_QUERY, sizeof(*scuq) + len);
	strlcpy(scuq->scuq_query, av[0], len);
}

int
slm_bml_prhdr(__unusedx struct psc_ctlmsghdr *mh,
    __unusedx const void *m)
{
	printf("%-16s %6s %-16s "
	    "%-15s %11s %9s\n",
	    "bmap-lease-fid", "bmapno", "io-system",
	    "client", "flags", "seqno");
	return(PSC_CTL_DISPLAY_WIDTH);
}

void
slm_bml_prdat(__unusedx const struct psc_ctlmsghdr *mh, const void *m)
{
	const struct slmctlmsg_bml *scbl = m;
	char *p, cli[PSCRPC_NIDSTR_SIZE];
	size_t n;

	strlcpy(cli, scbl->scbl_client, sizeof(cli));
	p = strchr(cli, '@');
	if (p)
		*p = '\0';
	p = cli;
	n = strcspn(p, "-");
	if (n)
		p += n + 1;

	printf("%016"SLPRIxFID" %6u "
	    "%-16.16s %-15s "
	    "%c%c%c%c%c%c%c%c%c%c "
	    "%9"PRIu64"\n",
	    scbl->scbl_fg.fg_fid, scbl->scbl_bno,
	    scbl->scbl_resname, p,
	    scbl->scbl_flags & BML_READ		? 'R' : '-',
	    scbl->scbl_flags & BML_WRITE	? 'W' : '-',
	    scbl->scbl_flags & BML_DIO		? 'D' : '-',
	    scbl->scbl_flags & BML_DIOCB	? 'C' : '-',
	    scbl->scbl_flags & BML_TIMEOQ	? 'T' : '-',
	    scbl->scbl_flags & BML_BMI		? 'B' : '-',
	    scbl->scbl_flags & BML_RECOVER	? 'V' : '-',
	    scbl->scbl_flags & BML_CHAIN	? 'N' : '-',
	    scbl->scbl_flags & BML_FREEING	? 'F' : '-',
	    scbl->scbl_flags & BML_RECOVERFAIL	? 'L' : '-',
	    scbl->scbl_seq);
}

struct psc_ctlshow_ent psc_ctlshow_tab[] = {
	PSC_CTLSHOW_DEFS,
	{ "bmaps",		packshow_bmaps },
	{ "bml",		packshow_bml },
	{ "connections",	packshow_conns },
	{ "fcmhs",		packshow_fcmhs },
	{ "replqueued",		packshow_replqueued },
	{ "statfs",		packshow_statfs },

	/* aliases */
	{ "conns",		packshow_conns },
	{ "fidcache",		packshow_fcmhs },
	{ "files",		packshow_fcmhs }
};

struct psc_ctlmsg_prfmt psc_ctlmsg_prfmts[] = {
	PSC_CTLMSG_PRFMT_DEFS,
	{ sl_bmap_prhdr,	sl_bmap_prdat,		sizeof(struct slctlmsg_bmap),		NULL },
	{ sl_conn_prhdr,	sl_conn_prdat,		sizeof(struct slctlmsg_conn),		NULL },
	{ sl_fcmh_prhdr,	sl_fcmh_prdat,		sizeof(struct slctlmsg_fcmh),		NULL },
	{ slm_replqueued_prhdr,	slm_replqueued_prdat,	sizeof(struct slmctlmsg_replqueued),	NULL },
	{ slm_statfs_prhdr,	slm_statfs_prdat,	sizeof(struct slmctlmsg_statfs),	NULL },
	{ NULL,			NULL,			0,					NULL },
	{ slm_bml_prhdr,	slm_bml_prdat,		sizeof(struct slmctlmsg_bml),		NULL }
};

struct psc_ctlcmd_req psc_ctlcmd_reqs[] = {
	{ "stop",		slmctlcmd_stop },
	{ "upsch-query",	slmctlcmd_upsch_query }
};

PFLCTL_CLI_DEFS;

const char *daemon_name = "slashd";

__dead void
usage(void)
{
	extern const char *__progname;

	fprintf(stderr,
	    "usage: %s [-HInV] [-p paramspec] [-S socket] [-s value] "
	    "[cmd arg ...]\n",
	    __progname);
	exit(1);
}

void
slmctl_show_version(void)
{
	fprintf(stderr, "%d\n", sl_stk_version);
}

struct psc_ctlopt opts[] = {
	{ 'H', PCOF_FLAG, &psc_ctl_noheader },
	{ 'I', PCOF_FLAG, &psc_ctl_inhuman },
	{ 'n', PCOF_FLAG, &psc_ctl_nodns },
	{ 'p', PCOF_FUNC, psc_ctlparse_param },
	{ 's', PCOF_FUNC, psc_ctlparse_show },
	{ 'V', PCOF_FLAG, slmctl_show_version },
};

int
main(int argc, char *argv[])
{
	pfl_init();
	sl_errno_init();
	pscthr_init(0, NULL, 0, "slmctl");

	psc_ctlcli_main(SL_PATH_SLMCTLSOCK, argc, argv, opts,
	    nitems(opts));
	exit(0);
}
